BedrockServer.cpp

```diff
@@ -2256,45 +2259,47 @@ void BedrockServer::handleSocket(Socket&& socket, bool fromControlPort, bool fro
                         // command that was already responded to in `buildCommandFromRequest` and we can move on to the
                         // next thing immediately.
                         mutex m;
-
-                        // Defer locking until we actually have to.
-                        unique_lock<mutex> lock(m, defer_lock);
                         condition_variable cv;
+                        atomic<bool> finished = false;
 
-                        function<void()> callback = [&m, &cv]() {
+                        function<void()> callback = [&m, &cv, &finished]() {
                             // Lock the mutex above (which will be locked by this thread while we're queuing), which waits
                             // for `handleSocket` to release it's lock (by calling `wait`), and then notify the waiting
                             // socket thread.
-                            lock_guard lock(m);
+                            lock_guard l(m);
+                            finished = true;
                             cv.notify_all();
                         };
 
-                        // Ok, none of above synchronization code gets called unless the command has a socket to respond
-                        // on.
+                        // Ok, none of above synchronization code gets called unless the command has a socket to respond on.
                         bool hasSocket = command->socket;
                         if (hasSocket) {
                             // Set the destructor callback for when the command finishes.
                             command->destructionCallback = &callback;
-
-                            // And lock the mutex so the command can't complete until we are in `wait` below.
-                            lock.lock();
                         }
 
-                        // Now we'll queue this command in one of two queues.
+                        // Now we queue or run this command.
                         auto _syncNodeCopy = atomic_load(&_syncNode);
                         if (_syncNodeCopy && _syncNodeCopy->getState() == SQLiteNode::STANDINGDOWN) {
                             _standDownQueue.push(move(command));
                         } else {
-                            SINFO("Queuing new '" << command->request.methodLine << "' command from local client, with "
-                                  << _commandQueue.size() << " commands already queued.");
-                            _commandQueue.push(move(command));
+                            // If this command is scheduled in the future, we can't just run it, we need to enqueue it to run at that point.
+                            // This functionality will go away as we remove the queues from bedrock, and so this can be removed at that time.
+                            if (command->request.calcU64("commandExecuteTime") > STimeNow()) {
```

Comment:
Where was this handled before?

Reply:
Here. Inserting it in the command queue handles it. [SScheduledPriorityQueue](https://github.com/Expensify/Bedrock/blob/main/libstuff/SScheduledPriorityQueue.h) doesn't dequeue commands that are scheduled in the future. They are enqueued behind any other commands that are scheduled earlier than them, and even if there are no commands in front of them, they're not dequeued until they're scheduled to run.

So, this maintains the behavior of pushing commands that are scheduled in the future (a feature we actually intend to remove eventually) into the command queue to be dequeued (and thus run) later.

This check is here to prevent them from being run immediately in `runCommand`, which is before they're scheduled.